sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala

```diff
@@ -76,6 +82,11 @@ class ObjectAggregationIterator(
    */
   processInputs()
 
+  TaskContext.get().addTaskCompletionListener[Unit](_ => {
+    // At the end of the task, update the task's spill size.
+    spillSize.set(TaskContext.get().taskMetrics().memoryBytesSpilled - spillSizeBefore)
```

Comment:
Just a comment; I feel it would be nicer if we could inspect object `hashMap` memory size (non-spilled case), too, but it looks difficult to do so, right?

Reply:
@maropu - this is a good point. Given the [map value is a general `InternalRow`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationMap.scala#L36), I feel it's hard to get an accurate metrics for memory size of this map. Hive aggregation uses a combination of [current JVM heap memory usage](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java#L872), and [estimation of map size](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java#L919-L920) to decide whether to spill. I think Hive's approach might be better. I am also tagging queries in our production environment with object hash aggregate fallback, and see how to improve them. Is it "memory-size-based object hash aggregation fallback" a good topic to discuss? cc @maropu and @cloud-fan thanks.

Reply:
Yea, the approach sounds good to me if possible. Anyway, it looks like we need to investigate how to estimate the size in a separate ticket.

Reply:
Thanks @maropu, created a followup JIRA - https://issues.apache.org/jira/browse/SPARK-34286 .